lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs

```diff
@@ -371,30 +453,72 @@ prop_runSelection_UTxO_notEnough (Small index) = monadicIO $ do
     let balanceSelected = view #balance selected
     let balanceLeftover = view #balance leftover
     assert $ balanceSelected == balanceRequested
     assert $ balanceLeftover == TokenBundle.empty
   where
     balanceRequested = view #balance index
 
+prop_runSelection_UTxO_extraSourceUsed
+    :: Maybe Coin
+    -> Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_extraSourceUsed extraSource (Small index) = do
+    let hasSomeAda = maybe False (/= TokenBundle.empty) almostEverything
+    cover 80 hasSomeAda "sometimes there are Ada" $ monadicIO $
+        case almostEverything of
+            Nothing ->
+                assert True
+            Just balanceRequested | hasSomeAda -> do
+                SelectionState {selected,leftover} <-
+                    run $ runSelection extraSource index balanceRequested
+                let balanceSelected = view #balance selected
+                let balanceLeftover = view #balance leftover
+                assert $ balanceLeftover == view #balance index
+                assert $ balanceSelected == TokenBundle.empty
+            Just balanceRequested -> do
+                SelectionState {selected} <-
+                    run $ runSelection extraSource index balanceRequested
+                let balanceSelected = view #balance selected
+                let coinSelected = TokenBundle.coin $
+                        addExtraSource extraSource balanceSelected
+                monitor $ counterexample $ unlines
+                    [ "balance selected: " <> pretty (Flat balanceSelected)
+                    ]
+                assert $ coinSelected >= TokenBundle.coin balanceRequested
+                assert $ balanceSelected /= TokenBundle.empty
+  where
+    almostEverything = TokenBundle.subtract
+        (view #balance index)
+        (TokenBundle.fromCoin (Coin 1))
+
 prop_runSelection_UTxO_moreThanEnough
-    :: Small UTxOIndex -> Property
-prop_runSelection_UTxO_moreThanEnough (Small index) = monadicIO $ do
+    :: Maybe Coin
+    -> Small UTxOIndex
+    -> Property
+prop_runSelection_UTxO_moreThanEnough extraSource (Small index) = monadicIO $ do
     SelectionState {selected, leftover} <-
-        run $ runSelection index balanceRequested
+        run $ runSelection extraSource index balanceRequested
     let balanceSelected = view #balance selected
     let balanceLeftover = view #balance leftover
     monitor $ cover 80
         (assetsRequested `Set.isProperSubsetOf` assetsAvailable)
         "assetsRequested ⊂ assetsAvailable"
     monitor $ cover 50 (Set.size assetsRequested >= 4)
         "size assetsRequested >= 4"
-    assert $ balanceRequested `leq` balanceSelected
+    monitor $ counterexample $ unlines
+        [ "balance available: " <> pretty (Flat balanceAvailable)
+        , "balance requested: " <> pretty (Flat balanceRequested)
+        , "balance selected:  " <> pretty (Flat balanceSelected)
+        , "balance leftover:  " <> pretty (Flat balanceLeftover)
```

Comment:
`Flat` works perfectly fine here, though `Nested` would give more concise output (fewer lines), which could make failures slightly easier to comprehend.

Reply:
I dislike `Nested` as I think it gives a wrong representation of the assets. I prefer to consider assets as a product between an AssetId and a Quantity. The PolicyId -> AssetName hierarchy is implementation detail which is irrelevant for reasoning about assets IMO.

Reply:
Fair enough! I don't have a strong feeling either way.

I've converted all the counter examples to use the `Flat` representation in 781ca4d0a5eb544724bfb6a9fae117c6bede0962.